<a href="https://colab.research.google.com/github/dboa9/automateLexWithTerraform-main23-10-23/blob/main/darija_english_trained_model24_9_24.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!huggingface-cli login


    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

    A token is already saved on your machine. Run `huggingface-cli whoami` to get more information or `huggingface-cli logout` if you want to log out.
    Setting a new token will erase the existing one.
    To login, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Enter your token (input will not be visible): 
Add token as git credential? (Y/n) y
Token is valid (permission: fineGr

In [ ]:
from google.colab import userdata
userdata.get('hf_token')

'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'

In [ ]:
from huggingface_hub import notebook_login
notebook_login()

https://www.youtube.com/watch?v=NwubVF3tXe4&list=PL_vpAsWR_TQVmh0gzE2-1E5gR3QqC5mhN

In [ ]:
# Transformers installation
! pip install transformers datasets
%pip install tokenized_datasets
%pip install -U pandas
%pip install --upgrade datasets
%pip install sacremoses
%pip install datasets transformers[sentencepiece] sacrebleu -q
%pip install torch
%pip install ipywidgets
%pip install --upgrade huggingface_hub
%pip install tensorflow
# To install from source instead of the last release, comment the command above and uncomment the following one.
! pip install git+https://github.com/huggingface/transformers.git

# Multilingual models for inference

There are several multilingual models in 🤗 Transformers, and their inference usage differs from monolingual models. Not *all* multilingual model usage is different though. Some models, like [bert-base-multilingual-uncased](https://huggingface.co/bert-base-multilingual-uncased), can be used just like a monolingual model. This guide will show you how to use multilingual models whose usage differs for inference.

## XLM

XLM has ten different checkpoints, only one of which is monolingual. The nine remaining model checkpoints can be split into two categories: the checkpoints that use language embeddings and those that don't.

### XLM with language embeddings

The following XLM models use language embeddings to specify the language used at inference:

- `xlm-mlm-ende-1024` (Masked language modeling, English-German)
- `xlm-mlm-enfr-1024` (Masked language modeling, English-French)
- `xlm-mlm-enro-1024` (Masked language modeling, English-Romanian)
- `xlm-mlm-xnli15-1024` (Masked language modeling, XNLI languages)
- `xlm-mlm-tlm-xnli15-1024` (Masked language modeling + translation, XNLI languages)
- `xlm-clm-enfr-1024` (Causal language modeling, English-French)
- `xlm-clm-ende-1024` (Causal language modeling, English-German)

Language embeddings are represented as a tensor of the same shape as the `input_ids` passed to the model. The values in these tensors depend on the language used and are identified by the tokenizer's `lang2id` and `id2lang` attributes.

In this example, load the `xlm-clm-enfr-1024` checkpoint (Causal language modeling, English-French):

In [ ]:
!pip install datasets transformers[sentencepiece] sacrebleu -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 kB 3.1 MB/s eta 0:00:00


In [ ]:
import os
import sys
import transformers
import tensorflow as tf
from datasets import load_dataset
from datasets import concatenate_datasets
from transformers import AutoTokenizer
from transformers import TFAutoModelForSeq2SeqLM, DataCollatorForSeq2Seq
from transformers import AdamWeightDecay
from transformers import AutoTokenizer, TFAutoModelForSeq2SeqLM

In [ ]:
import torch
from transformers import XLMTokenizer, XLMWithLMHeadModel

tokenizer = XLMTokenizer.from_pretrained("xlm-clm-enfr-1024")
model = XLMWithLMHeadModel.from_pretrained("xlm-clm-enfr-1024")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/131 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.45M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.01M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.02k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


pytorch_model.bin:   0%|          | 0.00/830M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

The `lang2id` attribute of the tokenizer displays this model's languages and their ids:

In [ ]:
model_checkpoint = "atlasia/Transliteration-Moroccan-Darija"

atlasia/Transliteration-Moroccan-Darija

Source:  https://huggingface.co/atlasia/Transliteration-Moroccan-Darija


The data set:

Source: https://huggingface.co/datasets/atlasia/darija_english

In [ ]:
from datasets import load_dataset

# List of subsets to load
subsets = ["web_data", "comments", "stories", "doda", "transliteration"]

# Dictionary to hold the loaded datasets
raw_datasets = {}

# Loop through each subset and load it separately
for subset in subsets:
    raw_datasets[subset] = load_dataset("atlasia/darija_english", subset)

# Now you can access each dataset by its subset name, for example:
print(raw_datasets["web_data"])
print(raw_datasets["comments"])


print(raw_datasets["transliteration"])
print(raw_datasets[ "stories"])
print(raw_datasets[ "doda"])

README.md:   0%|          | 0.00/1.09k [00:00<?, ?B/s]

atlasia_web_data.csv:   0%|          | 0.00/908k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/3000 [00:00<?, ? examples/s]

atlasia_comments.csv:   0%|          | 0.00/2.93M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/10000 [00:00<?, ? examples/s]

atlasia_stories.csv:   0%|          | 0.00/13.2M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/3000 [00:00<?, ? examples/s]

atlasia_doda.csv:   0%|          | 0.00/2.87M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/45103 [00:00<?, ? examples/s]

atlasia_atam.csv:   0%|          | 0.00/1.34M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/67186 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['english', 'darija', 'source'],
        num_rows: 3000
    })
})
DatasetDict({
    train: Dataset({
        features: ['id', 'english', 'darija', 'source'],
        num_rows: 10000
    })
})
DatasetDict({
    train: Dataset({
        features: ['darija_arabizi', 'darija_arabic'],
        num_rows: 67186
    })
})
DatasetDict({
    train: Dataset({
        features: ['ChapterName', 'ChapterLink', 'Author', 'Tags', 'darija', 'english', 'chunk_id', 'source'],
        num_rows: 3000
    })
})
DatasetDict({
    train: Dataset({
        features: ['id', 'darija', 'en'],
        num_rows: 45103
    })
})


In [ ]:
raw_datasets

{'web_data': DatasetDict({
     train: Dataset({
         features: ['english', 'darija', 'source'],
         num_rows: 3000
     })
 }),
 'comments': DatasetDict({
     train: Dataset({
         features: ['id', 'english', 'darija', 'source'],
         num_rows: 10000
     })
 }),
 'stories': DatasetDict({
     train: Dataset({
         features: ['ChapterName', 'ChapterLink', 'Author', 'Tags', 'darija', 'english', 'chunk_id', 'source'],
         num_rows: 3000
     })
 }),
 'doda': DatasetDict({
     train: Dataset({
         features: ['id', 'darija', 'en'],
         num_rows: 45103
     })
 }),
 'transliteration': DatasetDict({
     train: Dataset({
         features: ['darija_arabizi', 'darija_arabic'],
         num_rows: 67186
     })
 })}

In [ ]:
# Accessing the second row in the 'train' split of the 'web_data' subset
print(raw_datasets['web_data']['train'][1])

# Accessing the second row in the 'train' split of the 'comments' subset
print(raw_datasets['comments']['train'][1])

# Accessing the second row in the 'train' split of the 'stories' subset
print(raw_datasets['stories']['train'][1])

# Accessing the second row in the 'train' split of the 'doda' subset
print(raw_datasets['doda']['train'][1])

# Accessing the second row in the 'train' split of the 'transliteration' subset
print(raw_datasets['transliteration']['train'][1])

{'english': "I know him and I'm lying on the bed.", 'darija': 'تعرفيه وانا مجبد فوق السرير', 'source': 'WEB'}
{'id': 29080, 'english': 'Sister Halima, thank you for the recipe. Is it possible to add ground cocoa to it?', 'darija': 'ﺧﺘﻲ ﺣﻠﻴﻤﺔ ﺷﻜﺮﺍ ﻋﻠﻰ ﺍﻟﻮﺻﻔﺔ ﻭﺍﺵ ﻣﻤﻜﻦ ﻧﺪﻳﺮ ﻓﻴﻬﺎ ﻛﺎﻭﻛﺎﻭ ﻣﻄﺤﻮﻥ', 'source': 'COMMENTS'}
{'ChapterName': 'آلاء و المنحوس الجزء 76', 'ChapterLink': 'https://www.9esa.com/2018/08/aala2-walman7ous-76.html', 'Author': 'غزلان ش.', 'Tags': 2018, 'darija': '...اكيد عارفها واقفة...و كان عقلو كايقول ليه خليها تمدم تجاهلها قمعها قمعها قمعها...و لكن قبل ماتكمل التالتة كانت الدغمة في فمو...مايقدرش يكسر قلبها و هي جات حتال عندو...تبسمات و رجعات لبلاصتها...و كانت ديك الابتسامة كاافية ليه و مكفية...يتبع', 'english': "Sure, she was standing there... and his mind was telling him to let her serve, ignore her, suppress her, suppress her, suppress her... but before she could finish the third one, the morsel was in his mouth... he couldn't break her heart when she came all the way to h

In [ ]:
raw_datasets['transliteration']['train'][1]

{'darija_arabizi': 'cha3l', 'darija_arabic': 'شاعل'}

In [ ]:
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

tokenizer_config.json:   0%|          | 0.00/942 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.29M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/555 [00:00<?, ?B/s]

In [ ]:
tokenizer("helLo Preprocessing the data ")

{'input_ids': [7086, 3, 26, 3, 5877, 8838, 14, 342, 102, 18, 2378, 16, 19548], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [ ]:

# Tokenize the Darija sentence in Latin script
darija_latin = "bayna homa tay7awlo ib9aw mbrrdin"
darija_latin_tokens = tokenizer(darija_latin, return_tensors="tf")  # Use 'tf' for TensorFlow
print(f"Tokens for Darija (Latin script): {darija_latin_tokens}")

# Tokenize the Darija sentence in Arabic script
darija_arabic = "آلاء و المنحوس الجزء"
darija_arabic_tokens = tokenizer(darija_arabic, return_tensors="tf")
print(f"Tokens for Darija (Arabic script): {darija_arabic_tokens}")

# Tokenize the Darija sentence in Arabic script
darija_arabiz = "cha3l"
darija_arabiz_tokens = tokenizer(darija_arabiz, return_tensors="tf")
print(f"Tokens for Darija (Arabic script): {darija_arabiz_tokens}")

# Tokenize a  sentences (example in English)
sentence = "Hello. Preprocessing the data"

pair_tokens = tokenizer(sentence, return_tensors="tf")
print(f"Tokens for sentence : {pair_tokens}")

Tokens for Darija (Latin script): {'input_ids': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=
array([[5728, 6247, 2020, 3124, 1097, 1028,  551,  191,  478]],
      dtype=int32)>, 'token_type_ids': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=array([[0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)>, 'attention_mask': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=array([[1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=int32)>}
Tokens for Darija (Arabic script): {'input_ids': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=
array([[  43, 1763,   75,  105, 5528,   60,  105,  706,   42]],
      dtype=int32)>, 'token_type_ids': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=array([[0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)>, 'attention_mask': <tf.Tensor: shape=(1, 9), dtype=int32, numpy=array([[1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=int32)>}
Tokens for Darija (Arabic script): {'input_ids': <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[18992]], dtype=int32)>, 'token_type_ids': <tf.Tensor: shape=(1, 1), dtype=int32

In [ ]:


# # source_lang = "darija"
# # source_lang = "darija_arabizi"
# # target_lang = "english"
# # # Moroccan Darija tokenized?
# # # https://www.kaggle.com/code/mpwolke/moroccan-darija-tokenized
# # # https://huggingface.co/Kamel/t5-darija-summarization
# # # https://blog.gopenai.com/building-a-custom-mixture-of-experts-model-for-our-darija-from-tokenization-to-text-generation-b30ba49ae7cc

# # def preprocess_function(examples):
# #     inputs = [ex[source_lang] for ex in examples["train"]]
# #     inputs = [ex[source_lang] for ex in examples["train"]]
# #     targets = [ex[target_lang] for ex in examples ["train"]]
# #     inputs = [ex[source_lang] for ex in examples if source_lang in ex]
# #     targets = [ex[target_lang] for ex in examples if target_lang in ex]

# #       # Assuming 'darija' and 'english' fields are available in all datasets where applicable
# #     # inputs = [ex[source_lang] for ex in examples if source_lang in ex]
# #     # targets = [ex[target_lang] for ex in examples if target_lang in ex]
# #     # inputs = [ex['darija_arabizi'] for ex in examples]  # Adjust this based on your actual input feature
# #     # targets = [ex['english'] for ex in examples]  # Adjust this based on your actual target feature

# #     model_inputs = tokenizer(inputs, max_length=max_input_length, truncation=True)

# #     # Setup the tokenizer for targets
# #     with tokenizer.as_target_tokenizer():
# #         labels = tokenizer(targets, max_length=max_target_length, truncation=True)

# #     model_inputs["labels"] = labels["input_ids"]
# #     return model_inputs

# # -----------https://huggingface.co/atlasia/Transliteration-Moroccan-Darija
# from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
# from datasets import load_dataset

# max_input_length = 128
# max_target_length = 128
# # Load the tokenizer and model
# tokenizer = AutoTokenizer.from_pretrained("BounharAbdelaziz/Transliteration-Moroccan-Darija")
# model = AutoModelForSeq2SeqLM.from_pretrained("BounharAbdelaziz/Transliteration-Moroccan-Darija")
# def preprocess_function(examples):
#   # Define your Moroccan Darija Arabizi text
#   input_text = "cha3l"

#   # Tokenize the input text
#   input_tokens = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True)

#   # Perform transliteration
#   output_tokens = model.generate(**input_tokens)

#   # Decode the output tokens
#   output_text = tokenizer.decode(output_tokens[0], skip_special_tokens=True)

#   print("Transliteration:", output_text)

#   model_inputs = tokenizer(input_text, max_length=max_input_length, truncation=True)

#   #     # Setup the tokenizer for targets
#   with tokenizer.as_target_tokenizer():
#       labels = tokenizer(output_text, max_length=max_target_length, truncation=True)

#   model_inputs["labels"] = labels["input_ids"]

#   return model_inputs

# # Apply preprocess_function to each dataset split in the raw_datasets
# for key, dataset in raw_datasets.items():
#     tokenized_datasets = dataset['train'].map(preprocess_function, batched=True)
#     print(f"Processed {key} dataset")

from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
from datasets import load_dataset

# Tokenizer and model setup
max_input_length = 128
max_target_length = 128
tokenizer = AutoTokenizer.from_pretrained("BounharAbdelaziz/Transliteration-Moroccan-Darija")
model = AutoModelForSeq2SeqLM.from_pretrained("BounharAbdelaziz/Transliteration-Moroccan-Darija")

# Adjusting preprocess_function to handle different dataset fields

def preprocess_function(examples, dataset_type):
    # Define your input and target text based on dataset type
    if dataset_type == "transliteration":
        input_text = examples["darija_arabizi"]
        target_text = examples["darija_arabic"]
        language_id = 1  # Custom ID for Darija
    else:
        input_text = examples["darija"]
        target_text = examples["english"]
        language_id = 0  # Custom ID for English

    # Tokenize the input text
    model_inputs = tokenizer(input_text, max_length=max_input_length, truncation=True, padding="max_length")

    # Tokenize target text (e.g., the expected output)
    with tokenizer.as_target_tokenizer():
        labels = tokenizer(target_text, max_length=max_target_length, truncation=True, padding="max_length")

    # Add the language ID to the inputs
    # Convert input_ids to tensors and prepend language_id
    input_ids_with_lang = [[language_id] + ids for ids in model_inputs["input_ids"]]

    model_inputs["input_ids"] = input_ids_with_lang
    model_inputs["labels"] = labels["input_ids"]
    return model_inputs



# for key, dataset in raw_datasets.items():
#     print(f"Processing dataset: {key}")

#     # Apply map to the 'train' dataset in each split (e.g., raw_datasets['web_data']['train'])
#     tokenized_datasets = dataset["train"].map(lambda x: preprocess_function(x, key), batched=True)

#     print(f"Processed {key} dataset")


tokenizer_config.json:   0%|          | 0.00/942 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.29M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/555 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/4.55k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/127M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/121 [00:00<?, ?B/s]

In [ ]:
print(raw_datasets)

{'web_data': DatasetDict({
    train: Dataset({
        features: ['english', 'darija', 'source'],
        num_rows: 3000
    })
}), 'comments': DatasetDict({
    train: Dataset({
        features: ['id', 'english', 'darija', 'source'],
        num_rows: 10000
    })
}), 'stories': DatasetDict({
    train: Dataset({
        features: ['ChapterName', 'ChapterLink', 'Author', 'Tags', 'darija', 'english', 'chunk_id', 'source'],
        num_rows: 3000
    })
}), 'doda': DatasetDict({
    train: Dataset({
        features: ['id', 'darija', 'en'],
        num_rows: 45103
    })
}), 'transliteration': DatasetDict({
    train: Dataset({
        features: ['darija_arabizi', 'darija_arabic'],
        num_rows: 67186
    })
})}


In [ ]:
print(dataset["train"][:2])

{'english': ["I swear to God, you don't know me until you're about to be executed.", "I know him and I'm lying on the bed."], 'darija': ['ويقول لي راه والله وما عتارفتي تا غادي تعدم', 'تعرفيه وانا مجبد فوق السرير'], 'source': ['WEB', 'WEB']}


In [ ]:
# Process the first 2 samples from the 'train' split of the 'transliteration' dataset
processed_examples = preprocess_function(raw_datasets["transliteration"]["train"][:2], "transliteration")
print(processed_examples)


{'input_ids': [[1, 12034, 89, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [1, 18992, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], 'token_type_ids': [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [ ]:
for key, dataset in raw_datasets.items():
    print(f"Processing dataset: {key}")

    try:
        if "train" in dataset:
            tokenized_datasets = dataset["train"].map(lambda x: preprocess_function(x, key), batched=True)
            print(f"Processed {key} dataset")
        else:
            print(f"No train split found in {key} dataset.")
    except Exception as e:
        print(f"Error processing {key} dataset: {e}")


Processing dataset: web_data
Processed web_data dataset
Processing dataset: comments
Processed comments dataset
Processing dataset: stories
Processed stories dataset
Processing dataset: doda


Map:   0%|          | 0/45103 [00:00<?, ? examples/s]

Error processing doda dataset: 'english'
Processing dataset: transliteration


Map:   0%|          | 0/67186 [00:00<?, ? examples/s]

Processed transliteration dataset


In [ ]:
model = TFAutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

config.json:   0%|          | 0.00/4.55k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/127M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFEncoderDecoderModel.

All the weights of TFEncoderDecoderModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFEncoderDecoderModel for predictions without further training.


Next, create an example input:

In [ ]:
!pip install torch


In [ ]:
from datasets import DatasetDict
import torch
from torch.nn.utils.rnn import pad_sequence
from transformers import TrainingArguments, Trainer

# Assuming your dataset is in the variable `tokenized_datasets['train']`
input_ids = tokenized_datasets['train']['input_ids']

# Determine the number of rows (sequences)
num_rows = len(input_ids)

# Create a language embedding tensor filled with zeros, same size as input_ids
language_embedding = torch.zeros((num_rows, max(len(seq) for seq in input_ids)))  # For variable lengths

# Convert input_ids to a list of tensors for padding
input_ids_tensor = [torch.tensor(seq) for seq in input_ids]

# Pad input_ids and create the corresponding language embeddings
padded_input_ids = pad_sequence(input_ids_tensor, batch_first=True)
padded_language_embedding = torch.zeros_like(padded_input_ids)  # Same shape as padded input_ids

# Now, padded_language_embedding has the same shape as padded_input_ids
print("Padded Language Embedding shape:", padded_language_embedding.shape)
print("Padded Input IDs shape:", padded_input_ids.shape)

# Check the structure of the tokenized datasets
print(tokenized_datasets)

# Create train and validation splits from the training dataset
train_test_split = tokenized_datasets['train'].train_test_split(test_size=0.2)
tokenized_datasets = DatasetDict({
    'train': train_test_split['train'],
    'validation': train_test_split['test'],
})

# Set your training parameters
training_args = TrainingArguments(
    output_dir="./results",
    per_device_train_batch_size=batch_size,
    learning_rate=learning_rate,
    weight_decay=weight_decay,
    num_train_epochs=num_train_epochs,
    eval_strategy="epoch",  # Updated to eval_strategy
    save_strategy="epoch",
    logging_dir="./logs",
    # ---Autoomate push After trainin
    # push_to_hub=True,
    #  push_to_hub_model_id=f"deebee7/darija_english_translator",
    # report_to="tensorboard",

    # report_to="tensorboard",
)

# Initialize the Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    tokenizer=tokenizer,
)

# Now you can proceed with training


Padded Language Embedding shape: torch.Size([27518, 129])
Padded Input IDs shape: torch.Size([27518, 129])
DatasetDict({
    train: Dataset({
        features: ['darija_arabizi', 'darija_arabic', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 27518
    })
    validation: Dataset({
        features: ['darija_arabizi', 'darija_arabic', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 6880
    })
})


In [ ]:
data_collator = DataCollatorForSeq2Seq(tokenizer, model=model, return_tensors="tf")

Now you can pass the `input_ids` and language embedding to the model:

In [ ]:
generation_data_collator = DataCollatorForSeq2Seq(tokenizer, model=model, return_tensors="tf", pad_to_multiple_of=128)

In [ ]:
# def tokenize_function(examples):
#     # Directly access the string values for tokenization
#     return tokenizer(
#         text=examples['darija_arabizi'],  # Ensure this is a string
#         text_pair=examples['darija_arabic'],  # This should also be a string
#         padding='max_length',
#         truncation=True,
#         max_length=max_length,
#         return_tensors='pt'
#     )
# print(tokenized_datasets["train"][:5])  # Check the first 5 examples
# tokenized_datasets = tokenized_datasets.map(tokenize_function, batched=True)
# # Create the DataLoader
# train_dataloader = DataLoader(
#     tokenized_datasets["train"],
#     batch_size=batch_size,
#     shuffle=True,
#     collate_fn=data_collator,
# )

# # Print the first batch to verify
# for batch in train_dataloader:
#     print("Batch keys:", batch.keys())
#     print("Input IDs shape:", batch['input_ids'].shape)
#     print("Attention mask shape:", batch['attention_mask'].shape)
#     print("Labels shape:", batch['labels'].shape)
#     break  # Print only the first batch

def tokenize_function(examples):
    try:
        # Directly access the string values for tokenization
        tokenized = tokenizer(
            text=examples['darija_arabizi'],
            text_pair=examples['darija_arabic'],
            padding='max_length',
            truncation=True,
            max_length=max_length,
            return_tensors='pt'
        )
        # Optionally include labels if they exist
        if 'labels' in examples:
            tokenized['labels'] = tokenizer(examples['labels'], padding='max_length', truncation=True, max_length=max_length, return_tensors='pt')['input_ids']
        return tokenized
    except Exception as e:
        print(f"Error during tokenization: {e}")
        return {}


In [ ]:
from torch.utils.data import DataLoader

# Assuming tokenized_datasets["test"] is a PyTorch dataset
train_dataset = DataLoader(
    tokenized_datasets["train"],
    batch_size=batch_size,
    shuffle=True,
    collate_fn=data_collator,
)


The [run_generation.py](https://github.com/huggingface/transformers/tree/main/examples/pytorch/text-generation/run_generation.py) script can generate text with language embeddings using the `xlm-clm` checkpoints.

In [ ]:
print(tokenized_datasets)
from torch.utils.data import DataLoader

# Create DataLoader for training dataset
train_dataset = DataLoader(
    tokenized_datasets["train"],  # Assuming 'train' split exists
    batch_size=batch_size,
    shuffle=True,
    collate_fn=data_collator,
)

# Create DataLoader for validation dataset
validation_dataset = DataLoader(
    tokenized_datasets["validation"],  # Assuming 'validation' split exists
    batch_size=batch_size,
    shuffle=False,
    collate_fn=data_collator,
)


DatasetDict({
    train: Dataset({
        features: ['darija_arabizi', 'darija_arabic', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 22014
    })
    validation: Dataset({
        features: ['darija_arabizi', 'darija_arabic', 'input_ids', 'token_type_ids', 'attention_mask', 'labels'],
        num_rows: 5504
    })
})


In [ ]:
from torch.utils.data import DataLoader

# Assuming tokenized_datasets["validation"] is a valid Dataset object
generation_dataset = DataLoader(
    tokenized_datasets["validation"],
    batch_size=8,
    shuffle=False,
    collate_fn=generation_data_collator,
)


### XLM without language embeddings

In [ ]:
print(type(model))


<class 'transformers.models.encoder_decoder.modeling_encoder_decoder.EncoderDecoderModel'>


In [ ]:
from torch.utils.data import DataLoader

# Define a collate function (example)
def generation_data_collator(features):
    # Implement your logic to combine features into a batch
    return {
        'input_ids': torch.stack([f['input_ids'] for f in features]),
        'attention_mask': torch.stack([f['attention_mask'] for f in features]),
        'labels': torch.stack([f['labels'] for f in features]),
    }

# Create the DataLoader for the validation set
generation_dataset = DataLoader(
    tokenized_datasets["validation"],
    batch_size=8,
    shuffle=False,
    collate_fn=generation_data_collator,
)


In [ ]:
import torch
from torch.utils.data import DataLoader

# Set up device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Define the collate function
def generation_data_collator(features):
    input_ids = torch.tensor([f['input_ids'] for f in features])
    attention_mask = torch.tensor([f['attention_mask'] for f in features])
    labels = torch.tensor([f['labels'] for f in features])

    return {
        'input_ids': input_ids,
        'attention_mask': attention_mask,
        'labels': labels,
    }

# Create the DataLoader for the validation set
generation_dataset = DataLoader(
    tokenized_datasets["validation"],
    batch_size=8,
    shuffle=False,
    collate_fn=generation_data_collator,
)

# Generate outputs
model.eval()  # Set the model to evaluation mode

for batch in generation_dataset:
    # Move batch to the appropriate device
    batch = {k: v.to(device) for k, v in batch.items()}

    # Generate outputs
    with torch.no_grad():
        outputs = model.generate(input_ids=batch['input_ids'], attention_mask=batch['attention_mask'])

    # Process the outputs as needed (e.g., decode them)
    decoded_outputs = tokenizer.batch_decode(outputs, skip_special_tokens=True)
    print(decoded_outputs)


/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1258: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
  warnings.warn(


['جتيماعي ين', 'غانش عا اني يون', 'م ال سم يون', 'كان خلّص ين', 'كايعيّف', 'پي انو', 'مجلّج يو', 'نجح يت']
['كان غان كاتب يون', 'سوّ ينا ش', 'إست غان دوي', 'برّرات', 'غايستبعد', 'إ جزمو ا يون', 'ولا ه', 'نش ارجيو']
['غانخبّيو', 'بشوية', 'غاتّ كبّد د', 'غان كافحو', 'تبلّغو', 'عا ال دو ار', 'كاتع كسي ا كاتع ري', 'ستاغلّو ة']
['خو ر الح', 'غانس تي دعي', 'ماء كب ف', 'هزّو ين', 'كايعالجو', 'كاتع كب ر تخ ه ني', 'كاتر كب دا', 'أبدي']
['روماتيزم', 'غاتقودي', 'غات جفّو', 'غاتفكّري', 'كانقسمو', 'كايقلّلو', 'كاتعيّن ا', 'جداد']
['ماد يس يا ت', 'كايزيد', 'حلّ مّ', 'تدخّل ان مت', 'غاي تس يّرو عا', 'قتارح ت', 'وقفتو', 'عجب ة']
['إتبع يا', 'يسمعو', 'إر لّق ية', 'سد ت سراغنة', 'كات دوي ا تاص', 'حسب', 'كات وّ وّ شا ك', 'تسنّ يتو اب ن']
['إح هّ دو يون', 'غاي مزج ين', 'غاتشكّل ت', 'كات فرشخي ات ق', 'تز هرو جّي يون', 'تقة', 'إبلّغو', 'يعاود']
['تمنّاو', 'تساراحو', 'كانّ خافض دو يون', 'تعاونات', 'مشرو بين يون', 'قر يتي ط', 'ي طلّقو', 'طاليان']
['تواعدو', 'كايأمرو', 'غاتحبس', 'داي خ', 'ق بو تا', 'تخلّا', 'ف

In [ ]:
import torch
from torch import nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, Dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification

# Sample dataset class
class MyDataset(Dataset):
    def __init__(self, data):
        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]

# Initialize tokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

# Sample data for demonstration (replace with your actual data)
sample_data = [
    {
        "input_ids": tokenizer.encode("This is a sample input.", padding="max_length", max_length=10, truncation=True, return_tensors="pt").squeeze(0),  # Ensure correct shape
        "labels": torch.tensor(1)  # Ensure labels are tensors
    }
    for _ in range(100)
]

# Create dataset and DataLoader
dataset = MyDataset(sample_data)
dataloader = DataLoader(dataset, batch_size=8, shuffle=True)

# Create your model
model = AutoModelForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=2)

# Set your learning rate and weight decay
learning_rate = 5e-5  # Example learning rate
weight_decay = 0.01   # Example weight decay

# Set up the optimizer
optimizer = AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)

# Training loop
num_epochs = 3  # Example number of epochs
for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    for batch in dataloader:  # Iterate over the DataLoader
        optimizer.zero_grad()  # Zero the gradients

        # Forward pass - ensure the batch contains the right keys
        input_ids = batch['input_ids']
        labels = batch['labels']

        # Ensure input_ids has the correct shape (batch_size, sequence_length)
        # input_ids should already have shape (batch_size, sequence_length) due to DataLoader

        # Forward pass
        outputs = model(input_ids=input_ids, labels=labels)

        # Calculate loss
        loss = outputs.loss  # Use the loss from the model's outputs
        loss.backward()  # Backward pass
        optimizer.step()  # Update the weights

    print(f"Epoch {epoch + 1}/{num_epochs}, Loss: {loss.item()}")  # Print loss for each epoch


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
We strongly recommend passing in an `attention_mask` since your input_ids may be padded. See https://huggingface.co/docs/transformers/troubleshooting#incorrect-output-when-padding-tokens-arent-masked.


Epoch 1/3, Loss: 0.008678206242620945
Epoch 2/3, Loss: 0.002046594861894846
Epoch 3/3, Loss: 0.0010736914118751884


In [ ]:
import torch
from torch import nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, Dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification

# Sample dataset class
class MyDataset(Dataset):
    def __init__(self, data):
        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]

# Initialize tokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

# Sample data for demonstration (replace with your actual data)
train_data = [
    {
        "input_ids": tokenizer.encode("This is a sample input.", padding="max_length", max_length=10, truncation=True, return_tensors="pt").squeeze(0),
        "labels": torch.tensor(1)  # Ensure labels are tensors
    }
    for _ in range(100)
]

validation_data = [
    {
        "input_ids": tokenizer.encode("This is a validation input.", padding="max_length", max_length=10, truncation=True, return_tensors="pt").squeeze(0),
        "labels": torch.tensor(0)
    }
    for _ in range(20)
]

# Create datasets and DataLoaders
train_dataset = MyDataset(train_data)
train_dataloader = DataLoader(train_dataset, batch_size=8, shuffle=True)

validation_dataset = MyDataset(validation_data)
validation_dataloader = DataLoader(validation_dataset, batch_size=8, shuffle=False)

# Create your model
model = AutoModelForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=2)

# Set your learning rate and weight decay
learning_rate = 5e-5  # Example learning rate
weight_decay = 0.01   # Example weight decay

# Set up the optimizer
optimizer = AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)

# Training and Validation Loop
num_epochs = 1  # Example number of epochs
for epoch in range(num_epochs):
    # Training Phase
    model.train()  # Set the model to training mode
    total_loss = 0
    for batch in train_dataloader:  # Iterate over the training DataLoader
        optimizer.zero_grad()  # Zero the gradients

        # Forward pass
        input_ids = batch['input_ids']
        labels = batch['labels']

        outputs = model(input_ids=input_ids, labels=labels)

        # Calculate loss
        loss = outputs.loss  # Use the loss from the model's outputs
        total_loss += loss.item()

        loss.backward()  # Backward pass
        optimizer.step()  # Update the weights

    avg_train_loss = total_loss / len(train_dataloader)
    print(f"Epoch {epoch + 1}/{num_epochs}, Average Training Loss: {avg_train_loss:.4f}")

    # Validation Phase
    model.eval()  # Set the model to evaluation mode
    total_val_loss = 0
    with torch.no_grad():  # No need to track gradients during validation
        for batch in validation_dataloader:
            input_ids = batch['input_ids']
            labels = batch['labels']

            outputs = model(input_ids=input_ids, labels=labels)
            loss = outputs.loss  # Calculate validation loss
            total_val_loss += loss.item()

    avg_val_loss = total_val_loss / len(validation_dataloader)
    print(f"Average Validation Loss: {avg_val_loss:.4f}")


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch 1/1, Average Training Loss: 0.2865
Average Validation Loss: 3.2442


In [ ]:
import torch
from torch import nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, Dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification

# Sample dataset class
class MyDataset(Dataset):
    def __init__(self, data):
        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx]

# Initialize tokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

# Sample data for demonstration (replace with your actual data)
train_data = [
    tokenizer("This is a sample input.", padding="max_length", max_length=10, truncation=True, return_tensors="pt")
    for _ in range(100)
]
train_data = [{"input_ids": item["input_ids"].squeeze(0), "attention_mask": item["attention_mask"].squeeze(0), "labels": torch.tensor(1)} for item in train_data]

validation_data = [
    tokenizer("This is a validation input.", padding="max_length", max_length=10, truncation=True, return_tensors="pt")
    for _ in range(20)
]
validation_data = [{"input_ids": item["input_ids"].squeeze(0), "attention_mask": item["attention_mask"].squeeze(0), "labels": torch.tensor(0)} for item in validation_data]

# Create datasets and DataLoaders
train_dataset = MyDataset(train_data)
train_dataloader = DataLoader(train_dataset, batch_size=8, shuffle=True)

validation_dataset = MyDataset(validation_data)
validation_dataloader = DataLoader(validation_dataset, batch_size=8, shuffle=False)

# Create your model
model = AutoModelForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=2)

# Set your learning rate and weight decay
learning_rate = 5e-5  # Example learning rate
weight_decay = 0.01   # Example weight decay

# Set up the optimizer
optimizer = AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)

# Training and Validation Loop
num_epochs = 1  # Example number of epochs
for epoch in range(num_epochs):
    # Training Phase
    model.train()  # Set the model to training mode
    total_loss = 0
    for batch in train_dataloader:  # Iterate over the training DataLoader
        optimizer.zero_grad()  # Zero the gradients

        # Forward pass
        input_ids = batch['input_ids']
        attention_mask = batch['attention_mask']
        labels = batch['labels']

        outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)

        # Calculate loss
        loss = outputs.loss  # Use the loss from the model's outputs
        total_loss += loss.item()

        loss.backward()  # Backward pass
        optimizer.step()  # Update the weights

    avg_train_loss = total_loss / len(train_dataloader)
    print(f"Epoch {epoch + 1}/{num_epochs}, Average Training Loss: {avg_train_loss:.4f}")

    # Validation Phase
    model.eval()  # Set the model to evaluation mode
    total_val_loss = 0
    with torch.no_grad():  # No need to track gradients during validation
        for batch in validation_dataloader:
            input_ids = batch['input_ids']
            attention_mask = batch['attention_mask']
            labels = batch['labels']

            outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss  # Calculate validation loss
            total_val_loss += loss.item()

    avg_val_loss = total_val_loss / len(validation_dataloader)
    print(f"Average Validation Loss: {avg_val_loss:.4f}")


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch 1/1, Average Training Loss: 0.1453
Average Validation Loss: 4.4507


In [ ]:
model.save_pretrained("tf_model/")

In [ ]:
from transformers import AutoTokenizer, TFAutoModelForSeq2SeqLM

# Specify your model checkpoint
model_checkpoint = "t5-base"  # Change to the appropriate model name or path

# Initialize the tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

# Load the model
model = TFAutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)


config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFT5ForConditionalGeneration.

All the weights of TFT5ForConditionalGeneration were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFT5ForConditionalGeneration for predictions without further training.


In [ ]:
input_text  = "My name Bappy. My youtube channel name is DSwithBappy"

tokenized = tokenizer([input_text], return_tensors='np')
out = model.generate(**tokenized, max_length=128)
print(out)

tf.Tensor(
[[    0 32099  2659  9632     5   499    25  9863  4245   564    19     3
   3592  4065   279     9  9632     5   499    25  9863  4245   564    19
      3  3592  4065   279     9  9632     5   499    25  9863  4245   564
     19     3  3592  4065   279     9  9632     5   499    25  9863  4245
    564    19     3  3592  4065   279     9  9632     5   499    25  9863
   4245   564    19     3  3592  4065   279     9  9632     5   499    25
   9863  4245   564    19     3  3592  4065   279     9  9632     5   499
     25  9863  4245   564    19     3  3592  4065   279     9  9632     5
    499    25  9863  4245   564    19     3  3592  4065   279     9  9632
      5   499    25  9863  4245   564     1]], shape=(1, 115), dtype=int32)


In [ ]:
with tokenizer.as_target_tokenizer():
    print(tokenizer.decode(out[0], skip_special_tokens=True))

Bappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name


/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:4126: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your labels by using the argument `text_target` of the regular `__call__` method (either in the same call as your input texts if you use the same keyword arguments, or in a separate call.
  warnings.warn(


In [ ]:
out = model.generate(**tokenized, max_length=128, num_beams=5, do_sample=True, temperature=0.7)
input_text = "My name Bappy. My youtube channel name is DSwithBappy"
target_text = "This is an example target."

# Tokenizing inputs and targets together
tokenized = tokenizer([input_text], text_target=[target_text], return_tensors='np')

out = model.generate(**tokenized, max_length=128, num_beams=5, do_sample=True, temperature=0.7)
print(out)


tf.Tensor(
[[    0 32099   499   564    19  2659  9632     5   499    25  9863  4245
    564    19     3  3592  4065   279     9  9632     5   499    25  9863
   4245   564    19     3  3592  4065   279     9  9632     5   499    25
   9863  4245   564    19     3  3592  4065   279     9  9632     5   499
     25  9863  4245   564    19     3  3592  4065   279     9  9632     5
    499    25  9863  4245   564    19     3  3592  4065   279     9  9632
      5   499    25  9863  4245   564    19     3  3592  4065   279     9
   9632     5   499    25  9863  4245   564    19     3  3592  4065   279
      9  9632     5   499    25  9863  4245   564    19     3  3592  4065
    279     9  9632     5     1     0     0]], shape=(1, 115), dtype=int32)


In [ ]:
with tokenizer.as_target_tokenizer():
    print(tokenizer.decode(out[0], skip_special_tokens=True))

My name is Bappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy. My youtube channel name is DSwithBappy.


In [ ]:
input_text = "Tell me about myself. My name is Bappy, and my YouTube channel is called DSwithBappy."
tokenized = tokenizer([input_text], return_tensors='np')

out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=0.7,
    top_k=50,
    top_p=0.95
)

# Decoding the generated output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)
print(decoded_output)


Your name is Bappy.


In [ ]:
from transformers import Trainer, TrainingArguments, AutoTokenizer, AutoModelForSeq2SeqLM
from datasets import Dataset

# Load the model and tokenizer
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# Example training dataset
train_data = {
    'input_texts': [
        "Translate to Darija: My name is Bappy.",
        "Translate to Darija: I love programming."
    ],
    'target_texts': [
        "سميتي بابي.",
        "كنحب البرمجة."
    ]
}

# Example validation dataset
val_data = {
    'input_texts': [
        "Translate to Darija: This is a test.",
        "Translate to Darija: Machine learning is fun."
    ],
    'target_texts': [
        "هاد شي اختبار.",
        "التعلم الآلي ممتع."
    ]
}

# Create Datasets
train_dataset = Dataset.from_dict(train_data)
val_dataset = Dataset.from_dict(val_data)

# Tokenization function
def tokenize_function(examples):
    model_inputs = tokenizer(examples['input_texts'], padding="max_length", truncation=True, max_length=128)
    labels = tokenizer(examples['target_texts'], padding="max_length", truncation=True, max_length=128, return_tensors="pt")["input_ids"]
    model_inputs["labels"] = labels
    return model_inputs

# Tokenize the datasets
tokenized_train_dataset = train_dataset.map(tokenize_function, batched=True)
tokenized_val_dataset = val_dataset.map(tokenize_function, batched=True)

# Training arguments
training_args = TrainingArguments(
    output_dir='./results',          # output directory
    num_train_epochs=3,              # total number of training epochs
    per_device_train_batch_size=4,   # batch size per device during training
    save_steps=10_000,               # save checkpoint every 10,000 steps
    save_total_limit=2,               # limit the total amount of checkpoints
    eval_strategy="epoch",            # evaluate every epoch
)

# Trainer setup
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_val_dataset,  # Provide the evaluation dataset
)

# Start training
trainer.train()


Map:   0%|          | 0/2 [00:00<?, ? examples/s]

Map:   0%|          | 0/2 [00:00<?, ? examples/s]

Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


TrainOutput(global_step=3, training_loss=12.460220336914062, metrics={'train_runtime': 351.5464, 'train_samples_per_second': 0.017, 'train_steps_per_second': 0.009, 'total_flos': 812654788608.0, 'train_loss': 12.460220336914062, 'epoch': 3.0})

In [ ]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

# Load a translation model
# Note: If you find a specific model for Darija, replace the model_checkpoint accordingly
model_checkpoint = "Helsinki-NLP/opus-mt-en-ar"  # Change this if you find a specific Darija model
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# Prepare the input text
input_text = " My name is Bappy, and my YouTube channel is called DSwithBappy."

# Tokenization
tokenized = tokenizer([input_text], return_tensors='pt')

# Generate output
out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=1.0,  # Experiment with this parameter
    top_k=50,
    top_p=0.95
)

# Decode the output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)

# Print the result
print(decoded_output)




tokenizer_config.json:   0%|          | 0.00/44.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.39k [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/801k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/917k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.12M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


pytorch_model.bin:   0%|          | 0.00/308M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

اسمي بابي، وقناتي على اليوتيوب تسمى DSwithBaby.


In [ ]:
from transformers import Trainer, TrainingArguments, AutoTokenizer, AutoModelForSeq2SeqLM
from datasets import Dataset

# Load the model and tokenizer
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"
# model_checkpoint = "atlasia/Transliteration-Moroccan-Darija"
# model_checkpoint = "Helsinki-NLP/opus-mt-en-ar"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# Example training dataset
train_data = {
    'input_texts': [
        " My name is Bappy.",
        "Translate to Darija: I love programming."
    ],
    'target_texts': [
        "سميتي بابي.",
        "كنحب البرمجة."
    ]
}
train_data_1 = {
    'input_texts': [
    "My name is Bappy, and my YouTube channel is called DSwithBappy."
        "I love programming."
    ],
    'target_texts': [
"سميتي بابي، و قناتي على يوتيوب سميتها DSwithBappy.",
       "كنحب البرمجة."
    ]
}
# Example validation dataset
val_data = {
    'input_texts': [
       " This is a test.",
        " Machine learning is fun."
    ],
    'target_texts': [
        "هاد شي اختبار.",
        "التعلم الآلي ممتع."
    ]
}

# Create Datasets
train_dataset = Dataset.from_dict(train_data)
train_dataset_1 = Dataset.from_dict(train_data)
val_dataset = Dataset.from_dict(val_data)

# Tokenization function
def tokenize_function(examples):
    model_inputs = tokenizer(examples['input_texts'], padding="max_length", truncation=True, max_length=128)
    labels = tokenizer(examples['target_texts'], padding="max_length", truncation=True, max_length=128, return_tensors="pt")["input_ids"]
    model_inputs["labels"] = labels
    return model_inputs

# Tokenize the datasets
tokenized_train_dataset = train_dataset.map(tokenize_function, batched=True)
tokenized_val_dataset = val_dataset.map(tokenize_function, batched=True)

# Training arguments
training_args = TrainingArguments(
    output_dir='./results',          # output directory
    num_train_epochs=3,              # total number of training epochs
    per_device_train_batch_size=4,   # batch size per device during training
    save_steps=10_000,               # save checkpoint every 10,000 steps
    save_total_limit=2,               # limit the total amount of checkpoints
    eval_strategy="epoch",            # evaluate every epoch
)

# Trainer setup
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_val_dataset,  # Provide the evaluation dataset
)

# Start training
trainer.train()


tokenizer_config.json:   0%|          | 0.00/824 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/806k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/916k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.21M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/416 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.12k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/957M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

Map:   0%|          | 0/2 [00:00<?, ? examples/s]

Map:   0%|          | 0/2 [00:00<?, ? examples/s]

Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


TrainOutput(global_step=3, training_loss=13.203268686930338, metrics={'train_runtime': 77.6308, 'train_samples_per_second': 0.077, 'train_steps_per_second': 0.039, 'total_flos': 812654788608.0, 'train_loss': 13.203268686930338, 'epoch': 3.0})

In [ ]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

# Load a translation model
# Note: If you find a specific model for Darija, replace the model_checkpoint accordingly
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"  # Change this if you find a specific Darija model
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint)

# Prepare the input text
input_text = " My name is Bappy, and my YouTube channel is called DSwithBappy."

# Tokenization
tokenized = tokenizer([input_text], return_tensors='pt')

# Generate output
out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=1.0,  # Experiment with this parameter
    top_k=50,
    top_p=0.95
)

# Decode the output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)

# Print the result
print(decoded_output)

# Post-processing (Optional): You can manually adjust common phrases to Darija here


سميتي بابي, و قناتي ديال يوتيوب سميت دسوا بابي.


In [ ]:
model.save_pretrained("tf_model/")

Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


In [ ]:
# Transformers installation
! pip install transformers datasets
!pip install -U pandas
!pip install --upgrade datasets
!pip install sacremoses
!pip install torch
!pip install datasets transformers[sentencepiece] sacrebleu -q
!pip install safetensors
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
from datasets import Dataset

# Load a translation model (adjust checkpoint as necessary)
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

# Load model and set precision type (FP16)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint).half()  # Use .half() for FP16


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 474.3/474.3 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 7.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.9/39.9 MB 21.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 10.2 MB/s eta 0:00:00
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 14.0.2
    Uninstalling pyarrow-14.0.2:
      Successfully uninstalled pyarrow-14.0.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pyarrow<15.0.0a0,>=14.0.1, but you have pyarrow 17.0.0 which is incompatible.
ibis-framework 8.0.0 requires pyarrow<16,>=2, but you have pyarrow 17.0.0 which is incompatible.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 2.1 MB

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/824 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/806k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/916k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.21M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/416 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.12k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/957M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

In [ ]:
# Prepare the input text
input_text = "My name is Bappy, and my YouTube channel is called DSwithBappy."
tokenized = tokenizer([input_text], return_tensors='pt')


In [ ]:
# Generate output
out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=1.0,
    top_k=50,
    top_p=0.95
)

# Decode the output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)
print(decoded_output)


سميتي بابي, و قناتي ديال يوتيوب سميت دسوي بابي.


In [ ]:
# Example training and validation datasets
train_data = {
    'input_texts': [
        "My name is Bappy.",
        "I love programming."
    ],
    'target_texts': [
        "سميتي بابي.",
        "كنحب البرمجة."
    ]
}
val_data = {
    'input_texts': [
        "This is a test.",
        "Machine learning is fun."
    ],
    'target_texts': [
        "هاد شي اختبار.",
        "التعلم الآلي ممتع."
    ]
}

# Create Datasets
train_dataset = Dataset.from_dict(train_data)
val_dataset = Dataset.from_dict(val_data)


In [ ]:
def tokenize_function(examples):
    model_inputs = tokenizer(examples['input_texts'], padding="max_length", truncation=True, max_length=128)
    labels = tokenizer(examples['target_texts'], padding="max_length", truncation=True, max_length=128)["input_ids"]
    model_inputs["labels"] = labels
    return model_inputs

# Tokenize datasets
tokenized_train_dataset = train_dataset.map(tokenize_function, batched=True)
tokenized_val_dataset = val_dataset.map(tokenize_function, batched=True)


Map:   0%|          | 0/2 [00:00<?, ? examples/s]

Map:   0%|          | 0/2 [00:00<?, ? examples/s]

In [ ]:
from transformers import Trainer, TrainingArguments

training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=4,
    save_steps=10_000,
    save_total_limit=2,
    eval_strategy="epoch",
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_val_dataset,
)

# Start training
trainer.train()


Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


TrainOutput(global_step=3, training_loss=12.958333333333334, metrics={'train_runtime': 2605.3328, 'train_samples_per_second': 0.002, 'train_steps_per_second': 0.001, 'total_flos': 812654788608.0, 'train_loss': 12.958333333333334, 'epoch': 3.0})

In [ ]:
import torch

# Save model weights in a standard format
torch.save(model.state_dict(), "tf_model/model.pth")  # Save as .pth file


In [ ]:
# Prepare the input text
input_text = "My name is Bappy, and my YouTube channel is called DSwithBappy."
tokenized = tokenizer([input_text], return_tensors='pt')


In [ ]:
# Import necessary libraries
from huggingface_hub import login
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import os
import torch
from google.colab import drive, files

# Mount Google Drive
drive.mount('/content/drive')

# Create a directory for saving the model in Google Drive
output_dir = '/content/drive/MyDrive/tf_model-25-9-24-pm'
os.makedirs(output_dir, exist_ok=True)

# Load a translation model
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint).half()  # Convert to FP16

# Prepare the input text
input_text = "How is my beautiful angel."

# Tokenization
tokenized = tokenizer([input_text], return_tensors='pt')

# Generate output
out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=1.0,
    top_k=50,
    top_p=0.95
)

# Decode the output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)

# Print the result
print(decoded_output)

# Save the model and tokenizer
model_save_dir = "tf_model/"
os.makedirs(model_save_dir, exist_ok=True)

# Save the model and tokenizer
model.save_pretrained(model_save_dir)
tokenizer.save_pretrained(model_save_dir)

# Save model weights as PyTorch file
torch.save(model.state_dict(), os.path.join(model_save_dir, "model.pth"))

# Save config and tokenizer config
model.config.save_pretrained(model_save_dir)
tokenizer.save_pretrained(model_save_dir)

# Zip the model directory and save it to Google Drive
!zip -r "{output_dir}/tf_model-25-9-24-pm.zip" "{model_save_dir}"

# Download the zip file to local machine
files.download(f"{output_dir}/tf_model-25-9-24-pm.zip")


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


كيدايرا لملائكة زوينة ديالي.


Some non-default generation parameters are set in the model config. These should go into a GenerationConfig file (https://huggingface.co/docs/transformers/generation_strategies#save-a-custom-decoding-strategy-with-your-model) instead. This warning will be raised to an exception in v4.41.
Non-default generation parameters: {'max_length': 512, 'num_beams': 4, 'bad_words_ids': [[61246]], 'forced_eos_token_id': 25897}


updating: tf_model/ (stored 0%)
updating: tf_model/special_tokens_map.json (deflated 73%)
updating: tf_model/vocab.json (deflated 82%)
updating: tf_model/tokenizer_config.json (deflated 67%)
updating: tf_model/model.pth (deflated 8%)
updating: tf_model/model.safetensors (deflated 8%)
updating: tf_model/target.spm (deflated 55%)
updating: tf_model/config.json (deflated 59%)
updating: tf_model/generation_config.json (deflated 42%)
updating: tf_model/source.spm (deflated 50%)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
# Import necessary libraries
from huggingface_hub import login
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import os
import torch
from google.colab import drive, files

# Mount Google Drive
# drive.mount('/content/drive')

# Create a directory for saving the model in Google Drive
# output_dir = '/content/drive/MyDrive/tf_model-25-9-24-pm'
# os.makedirs(output_dir, exist_ok=True)

# Load a translation model
model_checkpoint = "lachkarsalim/Helsinki-translation-English_Moroccan-Arabic"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForSeq2SeqLM.from_pretrained(model_checkpoint).half()  # Convert to FP16

# Prepare the input text
input_text = "How is my beautiful angel."

# Tokenization
tokenized = tokenizer([input_text], return_tensors='pt')

# Generate output
out = model.generate(
    **tokenized,
    max_length=128,
    do_sample=True,
    temperature=1.0,
    top_k=50,
    top_p=0.95
)

# Decode the output
decoded_output = tokenizer.decode(out[0], skip_special_tokens=True)

# Print the result
print(decoded_output)

# Save the model and tokenizer
# model_save_dir = "tf_model/"
# os.makedirs(model_save_dir, exist_ok=True)

# Save the model and tokenizer
# model.save_pretrained(model_save_dir)
# tokenizer.save_pretrained(model_save_dir)

# Save model weights as PyTorch file
# torch.save(model.state_dict(), os.path.join(model_save_dir, "model.pth"))

# Save config and tokenizer config
# model.config.save_pretrained(model_save_dir)
# tokenizer.save_pretrained(model_save_dir)

# Zip the model directory and save it to Google Drive
# !zip -r "{output_dir}/tf_model-25-9-24-pm.zip" "{model_save_dir}"

# Download the zip file to local machine
files.download(f"{output_dir}/tf_model-25-9-24-pm.zip")


كيدايرا لمالّي زوين.


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

The following XLM models do not require language embeddings during inference:

- `xlm-mlm-17-1280` (Masked language modeling, 17 languages)
- `xlm-mlm-100-1280` (Masked language modeling, 100 languages)

These models are used for generic sentence representations, unlike the previous XLM checkpoints.

## BERT

The following BERT models can be used for multilingual tasks:

- `bert-base-multilingual-uncased` (Masked language modeling + Next sentence prediction, 102 languages)
- `bert-base-multilingual-cased` (Masked language modeling + Next sentence prediction, 104 languages)

These models do not require language embeddings during inference. They should identify the language from the
context and infer accordingly.

## XLM-RoBERTa

The following XLM-RoBERTa models can be used for multilingual tasks:

- `xlm-roberta-base` (Masked language modeling, 100 languages)
- `xlm-roberta-large` (Masked language modeling, 100 languages)

XLM-RoBERTa was trained on 2.5TB of newly created and cleaned CommonCrawl data in 100 languages. It provides strong gains over previously released multilingual models like mBERT or XLM on downstream tasks like classification, sequence labeling, and question answering.

## M2M100

The following M2M100 models can be used for multilingual translation:

- `facebook/m2m100_418M` (Translation)
- `facebook/m2m100_1.2B` (Translation)

In this example, load the `facebook/m2m100_418M` checkpoint to translate from Chinese to English. You can set the source language in the tokenizer:

In [ ]:
from transformers import M2M100ForConditionalGeneration, M2M100Tokenizer

en_text = "Do not meddle in the affairs of wizards, for they are subtle and quick to anger."
chinese_text = "不要插手巫師的事務, 因為他們是微妙的, 很快就會發怒."

tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M", src_lang="zh")
model = M2M100ForConditionalGeneration.from_pretrained("facebook/m2m100_418M")

Tokenize the text:

In [ ]:
encoded_zh = tokenizer(chinese_text, return_tensors="pt")

M2M100 forces the target language id as the first generated token to translate to the target language. Set the `forced_bos_token_id` to `en` in the `generate` method to translate to English:

In [ ]:
generated_tokens = model.generate(**encoded_zh, forced_bos_token_id=tokenizer.get_lang_id("en"))
tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)

'Do not interfere with the matters of the witches, because they are delicate and will soon be angry.'

## MBart

The following MBart models can be used for multilingual translation:

- `facebook/mbart-large-50-one-to-many-mmt` (One-to-many multilingual machine translation, 50 languages)
- `facebook/mbart-large-50-many-to-many-mmt` (Many-to-many multilingual machine translation, 50 languages)
- `facebook/mbart-large-50-many-to-one-mmt` (Many-to-one multilingual machine translation, 50 languages)
- `facebook/mbart-large-50` (Multilingual translation, 50 languages)
- `facebook/mbart-large-cc25`

In this example, load the `facebook/mbart-large-50-many-to-many-mmt` checkpoint to translate Finnish to English. You can set the source language in the tokenizer:

In [ ]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

en_text = "Do not meddle in the affairs of wizards, for they are subtle and quick to anger."
fi_text = "Älä sekaannu velhojen asioihin, sillä ne ovat hienovaraisia ja nopeasti vihaisia."

tokenizer = AutoTokenizer.from_pretrained("facebook/mbart-large-50-many-to-many-mmt", src_lang="fi_FI")
model = AutoModelForSeq2SeqLM.from_pretrained("facebook/mbart-large-50-many-to-many-mmt")

Tokenize the text:

In [ ]:
encoded_en = tokenizer(en_text, return_tensors="pt")

MBart forces the target language id as the first generated token to translate to the target language. Set the `forced_bos_token_id` to `en` in the `generate` method to translate to English:

In [ ]:
generated_tokens = model.generate(**encoded_en, forced_bos_token_id=tokenizer.lang_code_to_id("en_XX"))
tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)

"Don't interfere with the wizard's affairs, because they are subtle, will soon get angry."

If you are using the `facebook/mbart-large-50-many-to-one-mmt` checkpoint, you don't need to force the target language id as the first generated token otherwise the usage is the same.

In [ ]:
 from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:


%cd /content/tf_model-27-9-24-pm

/content/tf_model


In [ ]:
!zip -r tf_model-25-7-24-pm /content/tf_model-27-9-24-pm

  adding: content/tf_model/ (stored 0%)
  adding: content/tf_model/model.safetensors (deflated 7%)
  adding: content/tf_model/config.json (deflated 59%)
  adding: content/tf_model/generation_config.json (deflated 42%)
